# Contents:
### 01. Importing Libraries
### 02. Explore the Data
### 03. Describe the Data
### 04. Missing Values
### 05. Duplicate Values
### 06. Remove Unnecessary Columns
### 07. Replace Null Values
### 08. Separate dataframe by 'Group' column classification
### 09. Remove 'Group' column
### 10. Remove rows with "United States" classification
### 11. Remove Sex Distinction from dataframe
### 12. Remove 'Sex' column
### 13. Check COVID-19 Outliers
### 14. Export wrangled dataframe

# 01. Importing Libraries

In [108]:
# import libraries
import pandas as pd
import numpy as np
import os

In [109]:
path = r'/Users/kirstinhelgeson/Desktop/Data Analytics Course/07-2023 Achievement 6 Project'

In [110]:
# import dataframe
df_covid_deaths = pd.read_csv(os.path.join(path,'02 Data','Original Data','Provisional_COVID-19_Deaths_by_Sex_and_Age.csv'), index_col = False)

# 02. Explore the Data

In [111]:
# check first 5 entries of original dataframe
df_covid_deaths.head()

Data As Of  Start Date    End Date     Group  Year  Month          State  \
0  07/12/2023  01/01/2020  07/08/2023  By Total   NaN    NaN  United States   
1  07/12/2023  01/01/2020  07/08/2023  By Total   NaN    NaN  United States   
2  07/12/2023  01/01/2020  07/08/2023  By Total   NaN    NaN  United States   
3  07/12/2023  01/01/2020  07/08/2023  By Total   NaN    NaN  United States   
4  07/12/2023  01/01/2020  07/08/2023  By Total   NaN    NaN  United States   

         Sex     Age Group  COVID-19 Deaths  Total Deaths  Pneumonia Deaths  \
0  All Sexes      All Ages        1137145.0    11669359.0         1128515.0   
1  All Sexes  Under 1 year            504.0       69060.0             996.0   
2  All Sexes    0-17 years           1660.0      123239.0            2799.0   
3  All Sexes     1-4 years            275.0       13366.0             644.0   
4  All Sexes    5-14 years            501.0       20708.0             781.0   

   Pneumonia and COVID-19 Deaths  Influenza Deaths  \
0                       565564.0           21951.0   
1                           89.0              62.0   
2                          413.0             501.0   
3                           64.0             173.0   
4                          141.0             218.0   

   Pneumonia, Influenza, or COVID-19 Deaths Footnote  
0                                 1719583.0      NaN  
1                                    1471.0      NaN  
2                                    4522.0      NaN  
3                                    1019.0      NaN  
4                                    1346.0      NaN

In [112]:
# check last 5 entries of original dataframe
df_covid_deaths.tail()

Data As Of  Start Date    End Date     Group    Year  Month  \
132187  07/12/2023  07/01/2023  07/08/2023  By Month  2023.0    7.0   
132188  07/12/2023  07/01/2023  07/08/2023  By Month  2023.0    7.0   
132189  07/12/2023  07/01/2023  07/08/2023  By Month  2023.0    7.0   
132190  07/12/2023  07/01/2023  07/08/2023  By Month  2023.0    7.0   
132191  07/12/2023  07/01/2023  07/08/2023  By Month  2023.0    7.0   

              State     Sex          Age Group  COVID-19 Deaths  Total Deaths  \
132187  Puerto Rico  Female        50-64 years              0.0          14.0   
132188  Puerto Rico  Female        55-64 years              0.0          13.0   
132189  Puerto Rico  Female        65-74 years              0.0          23.0   
132190  Puerto Rico  Female        75-84 years              0.0          30.0   
132191  Puerto Rico  Female  85 years and over              NaN          39.0   

        Pneumonia Deaths  Pneumonia and COVID-19 Deaths  Influenza Deaths  \
132187               NaN                            0.0               0.0   
132188               NaN                            0.0               0.0   
132189               NaN                            0.0               0.0   
132190              10.0                            0.0               NaN   
132191              13.0                            NaN               NaN   

        Pneumonia, Influenza, or COVID-19 Deaths  \
132187                                       NaN   
132188                                       NaN   
132189                                       NaN   
132190                                      11.0   
132191                                      14.0   

                                                 Footnote  
132187  One or more data cells have counts between 1-9...  
132188  One or more data cells have counts between 1-9...  
132189  One or more data cells have counts between 1-9...  
132190  One or more data cells have counts between 1-9...  
132191  One or more data cells have counts between 1-9...

In [113]:
# check shape of original dataframe
df_covid_deaths.shape

(132192, 16)

In [114]:
# check basic info on dataframe columns
df_covid_deaths.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132192 entries, 0 to 132191
Data columns (total 16 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Data As Of                                132192 non-null  object 
 1   Start Date                                132192 non-null  object 
 2   End Date                                  132192 non-null  object 
 3   Group                                     132192 non-null  object 
 4   Year                                      129438 non-null  float64
 5   Month                                     118422 non-null  float64
 6   State                                     132192 non-null  object 
 7   Sex                                       132192 non-null  object 
 8   Age Group                                 132192 non-null  object 
 9   COVID-19 Deaths                           95001 non-null   float64
 10  Total Deaths        

### Data types are all consistent, but there are null values in many columns that need to be further examined.

In [115]:
# check list of column names in original dataframe
df_covid_deaths.columns

Index(['Data As Of', 'Start Date', 'End Date', 'Group', 'Year', 'Month',
       'State', 'Sex', 'Age Group', 'COVID-19 Deaths', 'Total Deaths',
       'Pneumonia Deaths', 'Pneumonia and COVID-19 Deaths', 'Influenza Deaths',
       'Pneumonia, Influenza, or COVID-19 Deaths', 'Footnote'],
      dtype='object')

# 03. Describe the Data

In [116]:
df_covid_deaths.describe()

Year          Month  COVID-19 Deaths  Total Deaths  \
count  129438.000000  118422.000000     9.500100e+04  1.131680e+05   
mean     2021.340426       6.093023     3.218328e+02  2.815108e+03   
std         1.057432       3.388188     6.055202e+03  5.475333e+04   
min      2020.000000       1.000000     0.000000e+00  0.000000e+00   
25%      2020.000000       3.000000     0.000000e+00  4.100000e+01   
50%      2021.000000       6.000000     0.000000e+00  1.500000e+02   
75%      2022.000000       9.000000     5.300000e+01  6.530000e+02   
max      2023.000000      12.000000     1.137145e+06  1.166936e+07   

       Pneumonia Deaths  Pneumonia and COVID-19 Deaths  Influenza Deaths  \
count      8.963300e+04                   97227.000000     106324.000000   
mean       3.384600e+02                     157.212966          5.161356   
std        6.080367e+03                    3020.712406        111.881139   
min        0.000000e+00                       0.000000          0.000000   
25%        0.000000e+00                       0.000000          0.000000   
50%        1.700000e+01                       0.000000          0.000000   
75%        7.500000e+01                      22.000000          0.000000   
max        1.128515e+06                  565564.000000      21951.000000   

       Pneumonia, Influenza, or COVID-19 Deaths  
count                              9.025400e+04  
mean                               5.116239e+02  
std                                9.241420e+03  
min                                0.000000e+00  
25%                                0.000000e+00  
50%                                2.500000e+01  
75%                                1.090000e+02  
max                                1.719583e+06

### None of these values seem out of the ordinary, however it should be noted that this dataframe contains rows of data for the entire United States during all of 2020-2023 (By Total), rows of data per year (By Year) and rows of data for each month of each year (By Month) so the statistical values such as the mean may not refer to the actual average death count from 2020 to 2023. May consider separating these categories into separate dataframes to allow for better statistical analysis, especially by month.

# 04. Missing Values

In [117]:
# check null value count for each column
df_covid_deaths.isnull().sum()

Data As Of                                      0
Start Date                                      0
End Date                                        0
Group                                           0
Year                                         2754
Month                                       13770
State                                           0
Sex                                             0
Age Group                                       0
COVID-19 Deaths                             37191
Total Deaths                                19024
Pneumonia Deaths                            42559
Pneumonia and COVID-19 Deaths               34965
Influenza Deaths                            25868
Pneumonia, Influenza, or COVID-19 Deaths    41938
Footnote                                    39143
dtype: int64

### The 2754 null values in the 'Year' column are because it is not necessary to record the year for the "By Total" entries in this dataframe since this refers to all three years 2020-2023. The 13770 null values in the 'Month' column are for the same reason - there are "By Total" and "By Year" entries included in this dataframe that make the 'Month' column irrelevant. All of the null values for 'COVID-19 Deaths', 'Total Deaths', 'Pneumonia Deaths', 'Pneumonia and COVID-19 Deaths', 'Influenza Deaths', 'Pneumonia, Influenza, or COVID-19 Deaths' are due to the following reason noted in the 'Footnote' column: "One or more data cells have counts between 1-9 and have been suppressed in accordance with NCHS confidentiality standards." The Footnote column has null values where there were no necessary comments such as this.

### I think the missing values in the Death columns can be replaced with a random integer 1-9 because we know that at least one death occurred and can't be reported because of cofidentiality and privacy issues, and based on how they update this data every week the newer counts in particular are likely an underestimate of the total deaths as they catch up with incoming death certificate data, so I do not believe that the data will be skewed irreparably by doing this. It is worth noting that null values make up 28%, 14%, 32%, 26%, 20%, and 32% of values in the 'COVID-19 Deaths', 'Total Deaths', 'Pneumonia Deaths', 'Pneumonia and COVID-19 Deaths', 'Influenza Deaths', 'Pneumonia, Influenza, or COVID-19 Deaths' columns respectively.

# 05. Duplicate Values

In [118]:
# create subset containing any duplicate values in the dataframe
df_dups = df_covid_deaths[df_covid_deaths.duplicated()]

In [119]:
# print df_dups to observe duplicates
df_dups

Empty DataFrame
Columns: [Data As Of, Start Date, End Date, Group, Year, Month, State, Sex, Age Group, COVID-19 Deaths, Total Deaths, Pneumonia Deaths, Pneumonia and COVID-19 Deaths, Influenza Deaths, Pneumonia, Influenza, or COVID-19 Deaths, Footnote]
Index: []

### No duplicates found.

# 06. Remove Unnecessary Columns

In [120]:
# remove 'Data As Of' column because we can just take note that this data was all updated 7/12/23.
# remove 'Start Date' column because it doesn't give us any additional information that the 'Month' column does not.
# remove 'End Date' column because it doesn't give us any additional information that the 'Month' column does not.

df_covid_deaths = df_covid_deaths.drop(columns=['Data As Of','Start Date','End Date'])

In [121]:
# check new output of dataframe after removing 3 columns
df_covid_deaths.head()

Group  Year  Month          State        Sex     Age Group  \
0  By Total   NaN    NaN  United States  All Sexes      All Ages   
1  By Total   NaN    NaN  United States  All Sexes  Under 1 year   
2  By Total   NaN    NaN  United States  All Sexes    0-17 years   
3  By Total   NaN    NaN  United States  All Sexes     1-4 years   
4  By Total   NaN    NaN  United States  All Sexes    5-14 years   

   COVID-19 Deaths  Total Deaths  Pneumonia Deaths  \
0        1137145.0    11669359.0         1128515.0   
1            504.0       69060.0             996.0   
2           1660.0      123239.0            2799.0   
3            275.0       13366.0             644.0   
4            501.0       20708.0             781.0   

   Pneumonia and COVID-19 Deaths  Influenza Deaths  \
0                       565564.0           21951.0   
1                           89.0              62.0   
2                          413.0             501.0   
3                           64.0             173.0   
4                          141.0             218.0   

   Pneumonia, Influenza, or COVID-19 Deaths Footnote  
0                                 1719583.0      NaN  
1                                    1471.0      NaN  
2                                    4522.0      NaN  
3                                    1019.0      NaN  
4                                    1346.0      NaN

### Remember that the 'By Total' group sums all data from 1/1/20 to 7/8/23.

# 07. Replace Null Values

In [122]:
# create list of columns to replace null values
columns_to_fill = ['COVID-19 Deaths', 'Total Deaths', 'Pneumonia Deaths', 'Pneumonia and COVID-19 Deaths', 'Influenza Deaths', 'Pneumonia, Influenza, or COVID-19 Deaths']

In [123]:
# create new dataframe that has all null values replaced by random integer from 1 to 9 inclusive
df_covid_deaths_replaced = df_covid_deaths[columns_to_fill].fillna(np.random.randint(1,10))

In [124]:
# add the other columns that were not changed from original dataframe
df_covid_deaths_replaced = pd.concat([df_covid_deaths.drop(columns=columns_to_fill), df_covid_deaths_replaced], axis=1)

In [125]:
# check null values for dataframe that replaces all null values
df_covid_deaths_replaced.isnull().sum()

Group                                           0
Year                                         2754
Month                                       13770
State                                           0
Sex                                             0
Age Group                                       0
Footnote                                    39143
COVID-19 Deaths                                 0
Total Deaths                                    0
Pneumonia Deaths                                0
Pneumonia and COVID-19 Deaths                   0
Influenza Deaths                                0
Pneumonia, Influenza, or COVID-19 Deaths        0
dtype: int64

In [126]:
# describe new dataframe
df_covid_deaths_replaced.describe()

Year          Month  COVID-19 Deaths  Total Deaths  \
count  129438.000000  118422.000000     1.321920e+05  1.321920e+05   
mean     2021.340426       6.093023     2.324135e+02  2.410556e+03   
std         1.057432       3.388188     5.135205e+03  5.067011e+04   
min      2020.000000       1.000000     0.000000e+00  0.000000e+00   
25%      2020.000000       3.000000     0.000000e+00  1.900000e+01   
50%      2021.000000       6.000000     4.000000e+00  9.900000e+01   
75%      2022.000000       9.000000     2.500000e+01  5.080000e+02   
max      2023.000000      12.000000     1.137145e+06  1.166936e+07   

       Pneumonia Deaths  Pneumonia and COVID-19 Deaths  Influenza Deaths  \
count      1.321920e+05                  132192.000000     132192.000000   
mean       2.307812e+02                     116.687886          4.934096   
std        5.009240e+03                    2591.476979        100.340006   
min        0.000000e+00                       0.000000          0.000000   
25%        0.000000e+00                       0.000000          0.000000   
50%        4.000000e+00                       4.000000          0.000000   
75%        3.700000e+01                      11.000000          1.000000   
max        1.128515e+06                  565564.000000      21951.000000   

       Pneumonia, Influenza, or COVID-19 Deaths  
count                              1.321920e+05  
mean                               3.505799e+02  
std                                7.639701e+03  
min                                0.000000e+00  
25%                                4.000000e+00  
50%                                4.000000e+00  
75%                                5.300000e+01  
max                                1.719583e+06

In [127]:
# export this version of the dataframe just to have a copy of the dataframe with newly inputed random integers 1 - 9
df_covid_deaths_replaced.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'COVID_deaths_replaced.csv'))

# 08. Separate dataframe by 'Group' column classification

### I want a dataframe for each classification under the 'Group' column, 'By Total', 'By Year', 'By Month'. This way I can isolate the data specifically for each month from 2020-2023. It's not that useful to just see total deaths for all 3 years (the 'By Total' data) other than to show how many people were lost during the COVID-19 pandemic from 2020-2023. We would not be able to show any trends by looking at this data in conjunction with the data observed by month.

In [128]:
# create a new dataframe with only rows that contain 'By Month' classification
df_covid_deaths_month = df_covid_deaths_replaced[df_covid_deaths_replaced['Group'] == 'By Month']

In [129]:
# check output of new dataframe
df_covid_deaths_month.head()

Group    Year  Month          State        Sex     Age Group  \
13770  By Month  2020.0    1.0  United States  All Sexes      All Ages   
13771  By Month  2020.0    1.0  United States  All Sexes  Under 1 year   
13772  By Month  2020.0    1.0  United States  All Sexes    0-17 years   
13773  By Month  2020.0    1.0  United States  All Sexes     1-4 years   
13774  By Month  2020.0    1.0  United States  All Sexes    5-14 years   

      Footnote  COVID-19 Deaths  Total Deaths  Pneumonia Deaths  \
13770      NaN              6.0      264677.0           17909.0   
13771      NaN              0.0        1784.0              41.0   
13772      NaN              0.0        2966.0              90.0   
13773      NaN              0.0         315.0              22.0   
13774      NaN              0.0         471.0              21.0   

       Pneumonia and COVID-19 Deaths  Influenza Deaths  \
13770                            3.0            2125.0   
13771                            0.0               8.0   
13772                            0.0              63.0   
13773                            0.0              18.0   
13774                            0.0              29.0   

       Pneumonia, Influenza, or COVID-19 Deaths  
13770                                   20037.0  
13771                                      49.0  
13772                                     153.0  
13773                                      40.0  
13774                                      50.0

In [130]:
# check shape of new dataframe
df_covid_deaths_month.shape

(118422, 13)

In [131]:
# check descriptive data of new dataframe
df_covid_deaths_month.describe()

Year          Month  COVID-19 Deaths   Total Deaths  \
count  118422.000000  118422.000000    118422.000000  118422.000000   
mean     2021.325581       6.093023        86.624048     896.916696   
std         1.050385       3.388188       996.555214    7267.505508   
min      2020.000000       1.000000         0.000000       0.000000   
25%      2020.000000       3.000000         0.000000      16.000000   
50%      2021.000000       6.000000         4.000000      79.000000   
75%      2022.000000       9.000000        18.000000     368.000000   
max      2023.000000      12.000000    105565.000000  373641.000000   

       Pneumonia Deaths  Pneumonia and COVID-19 Deaths  Influenza Deaths  \
count     118422.000000                  118422.000000     118422.000000   
mean          86.039562                      43.547559          2.021685   
std          804.691216                     517.579319         27.578119   
min            0.000000                       0.000000          0.000000   
25%            0.000000                       0.000000          0.000000   
50%            4.000000                       2.000000          0.000000   
75%           28.000000                       4.000000          0.000000   
max        69849.000000                   55416.000000       4459.000000   

       Pneumonia, Influenza, or COVID-19 Deaths  
count                              118422.00000  
mean                                  130.60395  
std                                  1284.03818  
min                                     0.00000  
25%                                     3.00000  
50%                                     4.00000  
75%                                    40.00000  
max                                120079.00000

In [132]:
# check info of new dataframe
df_covid_deaths_month.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 118422 entries, 13770 to 132191
Data columns (total 13 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Group                                     118422 non-null  object 
 1   Year                                      118422 non-null  float64
 2   Month                                     118422 non-null  float64
 3   State                                     118422 non-null  object 
 4   Sex                                       118422 non-null  object 
 5   Age Group                                 118422 non-null  object 
 6   Footnote                                  84137 non-null   object 
 7   COVID-19 Deaths                           118422 non-null  float64
 8   Total Deaths                              118422 non-null  float64
 9   Pneumonia Deaths                          118422 non-null  float64
 10  Pneumonia and CO

# 09. Remove 'Group' column

In [133]:
# remove 'group' column now since it no longer provides any meaningful information
df_covid_deaths_month_2 = df_covid_deaths_month.drop('Group', axis=1)

In [134]:
# check output of new dataframe
df_covid_deaths_month_2

Year  Month          State        Sex          Age Group  \
13770   2020.0    1.0  United States  All Sexes           All Ages   
13771   2020.0    1.0  United States  All Sexes       Under 1 year   
13772   2020.0    1.0  United States  All Sexes         0-17 years   
13773   2020.0    1.0  United States  All Sexes          1-4 years   
13774   2020.0    1.0  United States  All Sexes         5-14 years   
...        ...    ...            ...        ...                ...   
132187  2023.0    7.0    Puerto Rico     Female        50-64 years   
132188  2023.0    7.0    Puerto Rico     Female        55-64 years   
132189  2023.0    7.0    Puerto Rico     Female        65-74 years   
132190  2023.0    7.0    Puerto Rico     Female        75-84 years   
132191  2023.0    7.0    Puerto Rico     Female  85 years and over   

                                                 Footnote  COVID-19 Deaths  \
13770                                                 NaN              6.0   
13771                                                 NaN              0.0   
13772                                                 NaN              0.0   
13773                                                 NaN              0.0   
13774                                                 NaN              0.0   
...                                                   ...              ...   
132187  One or more data cells have counts between 1-9...              0.0   
132188  One or more data cells have counts between 1-9...              0.0   
132189  One or more data cells have counts between 1-9...              0.0   
132190  One or more data cells have counts between 1-9...              0.0   
132191  One or more data cells have counts between 1-9...              4.0   

        Total Deaths  Pneumonia Deaths  Pneumonia and COVID-19 Deaths  \
13770       264677.0           17909.0                            3.0   
13771         1784.0              41.0                            0.0   
13772         2966.0              90.0                            0.0   
13773          315.0              22.0                            0.0   
13774          471.0              21.0                            0.0   
...              ...               ...                            ...   
132187          14.0               4.0                            0.0   
132188          13.0               4.0                            0.0   
132189          23.0               4.0                            0.0   
132190          30.0              10.0                            0.0   
132191          39.0              13.0                            4.0   

        Influenza Deaths  Pneumonia, Influenza, or COVID-19 Deaths  
13770             2125.0                                   20037.0  
13771                8.0                                      49.0  
13772               63.0                                     153.0  
13773               18.0                                      40.0  
13774               29.0                                      50.0  
...                  ...                                       ...  
132187               0.0                                       4.0  
132188               0.0                                       4.0  
132189               0.0                                       4.0  
132190               4.0                                      11.0  
132191               4.0                                      14.0  

[118422 rows x 12 columns]

In [135]:
# check shape of new dataframe
df_covid_deaths_month_2.shape

(118422, 12)

In [136]:
# check descriptive data of new dataframe
df_covid_deaths_month_2.describe()

Year          Month  COVID-19 Deaths   Total Deaths  \
count  118422.000000  118422.000000    118422.000000  118422.000000   
mean     2021.325581       6.093023        86.624048     896.916696   
std         1.050385       3.388188       996.555214    7267.505508   
min      2020.000000       1.000000         0.000000       0.000000   
25%      2020.000000       3.000000         0.000000      16.000000   
50%      2021.000000       6.000000         4.000000      79.000000   
75%      2022.000000       9.000000        18.000000     368.000000   
max      2023.000000      12.000000    105565.000000  373641.000000   

       Pneumonia Deaths  Pneumonia and COVID-19 Deaths  Influenza Deaths  \
count     118422.000000                  118422.000000     118422.000000   
mean          86.039562                      43.547559          2.021685   
std          804.691216                     517.579319         27.578119   
min            0.000000                       0.000000          0.000000   
25%            0.000000                       0.000000          0.000000   
50%            4.000000                       2.000000          0.000000   
75%           28.000000                       4.000000          0.000000   
max        69849.000000                   55416.000000       4459.000000   

       Pneumonia, Influenza, or COVID-19 Deaths  
count                              118422.00000  
mean                                  130.60395  
std                                  1284.03818  
min                                     0.00000  
25%                                     3.00000  
50%                                     4.00000  
75%                                    40.00000  
max                                120079.00000

In [137]:
# check info of new dataframe
df_covid_deaths_month_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 118422 entries, 13770 to 132191
Data columns (total 12 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Year                                      118422 non-null  float64
 1   Month                                     118422 non-null  float64
 2   State                                     118422 non-null  object 
 3   Sex                                       118422 non-null  object 
 4   Age Group                                 118422 non-null  object 
 5   Footnote                                  84137 non-null   object 
 6   COVID-19 Deaths                           118422 non-null  float64
 7   Total Deaths                              118422 non-null  float64
 8   Pneumonia Deaths                          118422 non-null  float64
 9   Pneumonia and COVID-19 Deaths             118422 non-null  float64
 10  Influenza Deaths

# 10. Remove rows with "United States" classification

### I want a dataframe that does not have extra values or redundant information. If I have all of the information on COVID deaths by month and by state, then I can add these values together to get US totals if I need to. Since we are going to attempt a geographical analysis though I need the information to be about States rather than the entire country.

In [138]:
# create a new dataframe with only rows that don't contain 'United States' classification
df_covid_deaths_month_3 = df_covid_deaths_month_2[df_covid_deaths_month_2['State'] != 'United States']

In [139]:
# check output of new dataframe
df_covid_deaths_month_3

Year  Month        State        Sex          Age Group  \
15963   2020.0    1.0      Alabama  All Sexes           All Ages   
15964   2020.0    1.0      Alabama  All Sexes       Under 1 year   
15965   2020.0    1.0      Alabama  All Sexes         0-17 years   
15966   2020.0    1.0      Alabama  All Sexes          1-4 years   
15967   2020.0    1.0      Alabama  All Sexes         5-14 years   
...        ...    ...          ...        ...                ...   
132187  2023.0    7.0  Puerto Rico     Female        50-64 years   
132188  2023.0    7.0  Puerto Rico     Female        55-64 years   
132189  2023.0    7.0  Puerto Rico     Female        65-74 years   
132190  2023.0    7.0  Puerto Rico     Female        75-84 years   
132191  2023.0    7.0  Puerto Rico     Female  85 years and over   

                                                 Footnote  COVID-19 Deaths  \
15963   One or more data cells have counts between 1-9...              4.0   
15964                                                 NaN              0.0   
15965   One or more data cells have counts between 1-9...              0.0   
15966   One or more data cells have counts between 1-9...              0.0   
15967   One or more data cells have counts between 1-9...              0.0   
...                                                   ...              ...   
132187  One or more data cells have counts between 1-9...              0.0   
132188  One or more data cells have counts between 1-9...              0.0   
132189  One or more data cells have counts between 1-9...              0.0   
132190  One or more data cells have counts between 1-9...              0.0   
132191  One or more data cells have counts between 1-9...              4.0   

        Total Deaths  Pneumonia Deaths  Pneumonia and COVID-19 Deaths  \
15963         4729.0             282.0                            0.0   
15964           31.0               0.0                            0.0   
15965           71.0               4.0                            0.0   
15966           12.0               4.0                            0.0   
15967           16.0               4.0                            0.0   
...              ...               ...                            ...   
132187          14.0               4.0                            0.0   
132188          13.0               4.0                            0.0   
132189          23.0               4.0                            0.0   
132190          30.0              10.0                            0.0   
132191          39.0              13.0                            4.0   

        Influenza Deaths  Pneumonia, Influenza, or COVID-19 Deaths  
15963               35.0                                     318.0  
15964                0.0                                       0.0  
15965                4.0                                       4.0  
15966                0.0                                       4.0  
15967                0.0                                       4.0  
...                  ...                                       ...  
132187               0.0                                       4.0  
132188               0.0                                       4.0  
132189               0.0                                       4.0  
132190               4.0                                      11.0  
132191               4.0                                      14.0  

[116229 rows x 12 columns]

In [140]:
# check shape of new dataframe
df_covid_deaths_month_3.shape

(116229, 12)

In [141]:
# check descriptive data of new dataframe
df_covid_deaths_month_3.describe()

Year          Month  COVID-19 Deaths   Total Deaths  \
count  116229.000000  116229.000000    116229.000000  116229.000000   
mean     2021.325581       6.093023        44.362139     459.145351   
std         1.050385       3.388188       228.902161    1356.939052   
min      2020.000000       1.000000         0.000000       0.000000   
25%      2020.000000       3.000000         0.000000      15.000000   
50%      2021.000000       6.000000         4.000000      75.000000   
75%      2022.000000       9.000000        17.000000     337.000000   
max      2023.000000      12.000000     20065.000000   48148.000000   

       Pneumonia Deaths  Pneumonia and COVID-19 Deaths  Influenza Deaths  \
count     116229.000000                  116229.000000     116229.000000   
mean          44.270612                      22.370673          1.188309   
std          175.608276                     126.363248          6.023689   
min            0.000000                       0.000000          0.000000   
25%            0.000000                       0.000000          0.000000   
50%            4.000000                       0.000000          0.000000   
75%           26.000000                       4.000000          0.000000   
max        13802.000000                   12191.000000        746.000000   

       Pneumonia, Influenza, or COVID-19 Deaths  
count                             116229.000000  
mean                                  67.004474  
std                                  277.155057  
min                                    0.000000  
25%                                    0.000000  
50%                                    4.000000  
75%                                   37.000000  
max                                21687.000000

In [142]:
# check info of new dataframe
df_covid_deaths_month_3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 116229 entries, 15963 to 132191
Data columns (total 12 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Year                                      116229 non-null  float64
 1   Month                                     116229 non-null  float64
 2   State                                     116229 non-null  object 
 3   Sex                                       116229 non-null  object 
 4   Age Group                                 116229 non-null  object 
 5   Footnote                                  84137 non-null   object 
 6   COVID-19 Deaths                           116229 non-null  float64
 7   Total Deaths                              116229 non-null  float64
 8   Pneumonia Deaths                          116229 non-null  float64
 9   Pneumonia and COVID-19 Deaths             116229 non-null  float64
 10  Influenza Deaths

# 11. Remove Sex Distinction from dataframe

### I want my dataset to merge with another dataset soon and the other dataset does not have sex distinction included. I'm also not specifically interested in COVID-19 deaths by gender so don't need this information at all, but I need to remove the Male/Female distinction so that I only have 'All Sexes' rows which will more accurately represent my data.

In [143]:
# create a new dataframe with only rows that contain 'All Sexes' classification
df_covid_deaths_month_4 = df_covid_deaths_month_3[df_covid_deaths_month_3['Sex'] == 'All Sexes']

In [144]:
# check output of new dataframe
df_covid_deaths_month_4

Year  Month        State        Sex          Age Group  \
15963   2020.0    1.0      Alabama  All Sexes           All Ages   
15964   2020.0    1.0      Alabama  All Sexes       Under 1 year   
15965   2020.0    1.0      Alabama  All Sexes         0-17 years   
15966   2020.0    1.0      Alabama  All Sexes          1-4 years   
15967   2020.0    1.0      Alabama  All Sexes         5-14 years   
...        ...    ...          ...        ...                ...   
132153  2023.0    7.0  Puerto Rico  All Sexes        50-64 years   
132154  2023.0    7.0  Puerto Rico  All Sexes        55-64 years   
132155  2023.0    7.0  Puerto Rico  All Sexes        65-74 years   
132156  2023.0    7.0  Puerto Rico  All Sexes        75-84 years   
132157  2023.0    7.0  Puerto Rico  All Sexes  85 years and over   

                                                 Footnote  COVID-19 Deaths  \
15963   One or more data cells have counts between 1-9...              4.0   
15964                                                 NaN              0.0   
15965   One or more data cells have counts between 1-9...              0.0   
15966   One or more data cells have counts between 1-9...              0.0   
15967   One or more data cells have counts between 1-9...              0.0   
...                                                   ...              ...   
132153  One or more data cells have counts between 1-9...              0.0   
132154  One or more data cells have counts between 1-9...              0.0   
132155  One or more data cells have counts between 1-9...              4.0   
132156  One or more data cells have counts between 1-9...              4.0   
132157  One or more data cells have counts between 1-9...              4.0   

        Total Deaths  Pneumonia Deaths  Pneumonia and COVID-19 Deaths  \
15963         4729.0             282.0                            0.0   
15964           31.0               0.0                            0.0   
15965           71.0               4.0                            0.0   
15966           12.0               4.0                            0.0   
15967           16.0               4.0                            0.0   
...              ...               ...                            ...   
132153          41.0               4.0                            0.0   
132154          34.0               4.0                            0.0   
132155          49.0              11.0                            4.0   
132156          79.0              26.0                            4.0   
132157          79.0              25.0                            4.0   

        Influenza Deaths  Pneumonia, Influenza, or COVID-19 Deaths  
15963               35.0                                     318.0  
15964                0.0                                       0.0  
15965                4.0                                       4.0  
15966                0.0                                       4.0  
15967                0.0                                       4.0  
...                  ...                                       ...  
132153               0.0                                       4.0  
132154               0.0                                       4.0  
132155               0.0                                      12.0  
132156               4.0                                      28.0  
132157               4.0                                      26.0  

[38743 rows x 12 columns]

In [145]:
# check shape of new dataframe
df_covid_deaths_month_4.shape

(38743, 12)

In [146]:
# check descriptive data of new dataframe
df_covid_deaths_month_4.describe()

Year         Month  COVID-19 Deaths  Total Deaths  \
count  38743.000000  38743.000000     38743.000000  38743.000000   
mean    2021.325581      6.093023        66.422321    688.735875   
std        1.050394      3.388217       321.511657   1901.526569   
min     2020.000000      1.000000         0.000000      0.000000   
25%     2020.000000      3.000000         0.000000     26.000000   
50%     2021.000000      6.000000         4.000000    122.000000   
75%     2022.000000      9.000000        26.000000    543.000000   
max     2023.000000     12.000000     20065.000000  48148.000000   

       Pneumonia Deaths  Pneumonia and COVID-19 Deaths  Influenza Deaths  \
count      38743.000000                   38743.000000      38743.000000   
mean          66.246238                      33.445190          1.671347   
std          246.312164                     177.343424          8.390965   
min            0.000000                       0.000000          0.000000   
25%            4.000000                       0.000000          0.000000   
50%            4.000000                       4.000000          0.000000   
75%           41.000000                      11.000000          0.000000   
max        13802.000000                   12191.000000        746.000000   

       Pneumonia, Influenza, or COVID-19 Deaths  
count                              38743.000000  
mean                                 100.359729  
std                                  389.029407  
min                                    0.000000  
25%                                    4.000000  
50%                                    4.000000  
75%                                   59.000000  
max                                21687.000000

In [147]:
# check info of new dataframe
df_covid_deaths_month_4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38743 entries, 15963 to 132157
Data columns (total 12 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Year                                      38743 non-null  float64
 1   Month                                     38743 non-null  float64
 2   State                                     38743 non-null  object 
 3   Sex                                       38743 non-null  object 
 4   Age Group                                 38743 non-null  object 
 5   Footnote                                  27653 non-null  object 
 6   COVID-19 Deaths                           38743 non-null  float64
 7   Total Deaths                              38743 non-null  float64
 8   Pneumonia Deaths                          38743 non-null  float64
 9   Pneumonia and COVID-19 Deaths             38743 non-null  float64
 10  Influenza Deaths             

# 12. Remove 'Sex' column

In [148]:
# remove 'Sex' column because it is no longer necessary or provides any unique information
df_covid_deaths_month_5 = df_covid_deaths_month_4.drop('Sex', axis=1)

In [149]:
# check output of new dataframe
df_covid_deaths_month_5

Year  Month        State          Age Group  \
15963   2020.0    1.0      Alabama           All Ages   
15964   2020.0    1.0      Alabama       Under 1 year   
15965   2020.0    1.0      Alabama         0-17 years   
15966   2020.0    1.0      Alabama          1-4 years   
15967   2020.0    1.0      Alabama         5-14 years   
...        ...    ...          ...                ...   
132153  2023.0    7.0  Puerto Rico        50-64 years   
132154  2023.0    7.0  Puerto Rico        55-64 years   
132155  2023.0    7.0  Puerto Rico        65-74 years   
132156  2023.0    7.0  Puerto Rico        75-84 years   
132157  2023.0    7.0  Puerto Rico  85 years and over   

                                                 Footnote  COVID-19 Deaths  \
15963   One or more data cells have counts between 1-9...              4.0   
15964                                                 NaN              0.0   
15965   One or more data cells have counts between 1-9...              0.0   
15966   One or more data cells have counts between 1-9...              0.0   
15967   One or more data cells have counts between 1-9...              0.0   
...                                                   ...              ...   
132153  One or more data cells have counts between 1-9...              0.0   
132154  One or more data cells have counts between 1-9...              0.0   
132155  One or more data cells have counts between 1-9...              4.0   
132156  One or more data cells have counts between 1-9...              4.0   
132157  One or more data cells have counts between 1-9...              4.0   

        Total Deaths  Pneumonia Deaths  Pneumonia and COVID-19 Deaths  \
15963         4729.0             282.0                            0.0   
15964           31.0               0.0                            0.0   
15965           71.0               4.0                            0.0   
15966           12.0               4.0                            0.0   
15967           16.0               4.0                            0.0   
...              ...               ...                            ...   
132153          41.0               4.0                            0.0   
132154          34.0               4.0                            0.0   
132155          49.0              11.0                            4.0   
132156          79.0              26.0                            4.0   
132157          79.0              25.0                            4.0   

        Influenza Deaths  Pneumonia, Influenza, or COVID-19 Deaths  
15963               35.0                                     318.0  
15964                0.0                                       0.0  
15965                4.0                                       4.0  
15966                0.0                                       4.0  
15967                0.0                                       4.0  
...                  ...                                       ...  
132153               0.0                                       4.0  
132154               0.0                                       4.0  
132155               0.0                                      12.0  
132156               4.0                                      28.0  
132157               4.0                                      26.0  

[38743 rows x 11 columns]

# 13. Check COVID-19 Outliers

In [150]:
# calculating quartiles
Q1 = df_covid_deaths_month_5['COVID-19 Deaths'].quantile(0.25)
Q2 = df_covid_deaths_month_5['COVID-19 Deaths'].quantile(0.50)
Q3 = df_covid_deaths_month_5['COVID-19 Deaths'].quantile(0.75)

In [151]:
# show calculated quartiles
print("Q1:", Q1)
print("Q2:", Q2)
print("Q3:", Q3)

Q1: 0.0
Q2: 4.0
Q3: 26.0


In [152]:
# calculate standard deviation
std_dev = df_covid_deaths_month_5['COVID-19 Deaths'].std()

In [153]:
# show standard deviation
std_dev

321.5116570180458

In [154]:
# calculate mean
mean = df_covid_deaths_month_5['COVID-19 Deaths'].mean()

In [155]:
# show mean
mean

66.42232145161707

In [156]:
# calculate two standard deviations above mean (upper bound)
upper_bound = mean + (2 * std_dev)

In [157]:
# show upper bound
upper_bound

709.4456354877087

In [158]:
# calculate two standard deviations below mean (lower bound)
lower_bound = mean - (2 * std_dev)

In [159]:
# show lower bound
lower_bound

-576.6009925844745

### The lower bound here is negative which is not sensical when talking about population deaths, so the lower bound is effectively 0.

In [160]:
# define outliers
COVID_outliers = df_covid_deaths_month_5[(df_covid_deaths_month_5['COVID-19 Deaths'] < lower_bound) | (df_covid_deaths_month_5['COVID-19 Deaths'] > upper_bound)]

In [161]:
# show resulting dataframe containing only COVID-19 Death outliers
COVID_outliers

Year  Month      State Age Group  \
16269   2020.0    7.0    Alabama  All Ages   
16320   2020.0    8.0    Alabama  All Ages   
16473   2020.0   11.0    Alabama  All Ages   
16524   2020.0   12.0    Alabama  All Ages   
16575   2021.0    1.0    Alabama  All Ages   
...        ...    ...        ...       ...   
126225  2021.0    1.0  Wisconsin  All Ages   
126684  2021.0   10.0  Wisconsin  All Ages   
126735  2021.0   11.0  Wisconsin  All Ages   
126786  2021.0   12.0  Wisconsin  All Ages   
126837  2022.0    1.0  Wisconsin  All Ages   

                                                 Footnote  COVID-19 Deaths  \
16269   One or more data cells have counts between 1-9...            879.0   
16320   One or more data cells have counts between 1-9...            964.0   
16473   One or more data cells have counts between 1-9...            768.0   
16524   One or more data cells have counts between 1-9...           1749.0   
16575   One or more data cells have counts between 1-9...           2415.0   
...                                                   ...              ...   
126225  One or more data cells have counts between 1-9...           1041.0   
126684  One or more data cells have counts between 1-9...            740.0   
126735                                                NaN            890.0   
126786  One or more data cells have counts between 1-9...           1341.0   
126837                                                NaN           1476.0   

        Total Deaths  Pneumonia Deaths  Pneumonia and COVID-19 Deaths  \
16269         5717.0             522.0                          279.0   
16320         5765.0             572.0                          334.0   
16473         5548.0             456.0                          232.0   
16524         7025.0             854.0                          581.0   
16575         7821.0            1164.0                          891.0   
...              ...               ...                            ...   
126225        5684.0             565.0                          368.0   
126684        5704.0             548.0                          333.0   
126735        5770.0             610.0                          380.0   
126786        6468.0             813.0                          569.0   
126837        6746.0             766.0                          550.0   

        Influenza Deaths  Pneumonia, Influenza, or COVID-19 Deaths  
16269                4.0                                    1124.0  
16320                4.0                                    1203.0  
16473                4.0                                     993.0  
16524                4.0                                    2026.0  
16575                4.0                                    2692.0  
...                  ...                                       ...  
126225               4.0                                    1239.0  
126684               4.0                                     956.0  
126735               0.0                                    1120.0  
126786               4.0                                    1589.0  
126837              16.0                                    1705.0  

[680 rows x 11 columns]

In [162]:
COVID_outliers.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 680 entries, 16269 to 126837
Data columns (total 11 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Year                                      680 non-null    float64
 1   Month                                     680 non-null    float64
 2   State                                     680 non-null    object 
 3   Age Group                                 680 non-null    object 
 4   Footnote                                  469 non-null    object 
 5   COVID-19 Deaths                           680 non-null    float64
 6   Total Deaths                              680 non-null    float64
 7   Pneumonia Deaths                          680 non-null    float64
 8   Pneumonia and COVID-19 Deaths             680 non-null    float64
 9   Influenza Deaths                          680 non-null    float64
 10  Pneumonia, Influenza, or COVID-

### There are 680 entries of total 38743 that are considered outliers for COVID-19 Death Counts. We will leave these outliers alone since this data is sourced and collected reliably and does not indicate any error in the data.

# 14. Export wrangled dataframe

In [163]:
df_covid_deaths_month_5.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'COVID_deaths_by_month_wrangled.csv'))